# ![](https://ga-dash.s3.amazonaws.com/production/assets/logo-9f88ae6c9c3871690e33280fcf557f33.png) Pandas for Exploratory Data Analysis I 
by [@josephofiowa](https://twitter.com/josephofiowa)

Pandas is the most prominent Python library for exploratory data analysis (EDA). The functions Pandas supports are integral to understanding, formatting, and preparing our data. Formally, we use Pandas to investigate, wrangle, munge, and clean our data. Pandas is the Swiss Army Knife of data manipulation!


We'll have two coding-heavy sessions on Pandas. In this one, we'll use Pandas to:
 - Read in a dataset
 - Investigate a dataset's integrity
 - Filter, sort, and manipulate a DataFrame's series

*An important message from our sponsor before going further:*

![](https://media.giphy.com/media/UtxwEhibdd5ss/giphy.gif)

## About the Dataset: Iowa Liquor

For today's Pandas exercises, we will be using a real dataset from the state of Iowa government on liquor sales. Due to state licensing laws, stores must report daily transactions of all alcohol they sell to the Iowa Department of Commerce's Alcoholic Beverages Division. The state of Iowa makes this data available for analysis -- and it is an excellent, structured dataset for our use!

Take a look at the data source [page](https://data.iowa.gov/Economy/Iowa-Liquor-Sales/m3tr-qhgy).


Let's take a closer look at the data dictionary, or what is included:
- **Invoice/Item Number** - Concatenated invoice and line number associated with the liquor order. This provides a unique identifier for the individual liquor products included in the store order
- **Date** - Date of order 
- **Store Number** - Unique number assigned to the store who ordered the liquor.
- **Store Name** - Name of store who ordered the liquor.
- **Address** - Address of the store that ordered the liquor
- **City** - City where the store who ordered the liquor is located
- **Zip Code** - Zip Code of where the store that ordered is located 
- **Store Location** - Location of store who ordered the liquor. The Address, City, State and Zip Code are geocoded to provide geographic coordinates. Accuracy of geocoding is dependent on how well the address is interpreted and the completeness of the reference data used.
- **County Number** - Iowa county number for the county where store who ordered the liquor is located
- **County** - County where the store who ordered the liquor is located
- **Category** - Category code associated with the liquor ordered
- **Category Names** - Category of the liquor ordered.
- **Vendor Number** - The vendor number of the company for the brand of liquor ordered
- **Vendor Name** - The vendor name of the company for the brand of liquor ordered
- **Item Name** - Item number for the individual liquor product ordered.
- **Item Description** - Description of the individual liquor product ordered.
- **Pack** - The number of bottles in a case for the liquor ordered
- **Bottle Volume (mL)** - Volume of each liquor bottle ordered in milliliters.
- **State Bottle Cost** - The amount that Alcoholic Beverages Division paid for each bottle of liquor ordered
- **State Bottle Retail** - The amount the store paid for each bottle of liquor ordered
- **Bottles Solde** - The number of bottles of liquor ordered by the store
- **Sale (Dollars)** - Total cost of liquor order (number of bottles multiplied by the state bottle retail)
- **Volume Sold (Liters)** - Total volume of liquor ordered in liters. (i.e. (Bottle Volume (ml) x Bottles Sold)/1,000)
- **Volume Sold (Gallons)** - Total volume of liquor ordered in gallons. (i.e. (Bottle Volume (ml) x Bottles Sold)/3785.411784)


### Our Modified Iowa Liquor Dataset

Because the full dataset (of all liquor sales from 2012 to present) is greater than 13 million rows (13,948,103+ at the time of writing), **we will work with a modified dataset.**

Our modified dataset has a few key changes:
- Only sales from May 2017 and May 2018 are present
- A number of values have been deliberately deleted (to practice working with missing data!)


## Importing Pandas

To import a library, we simply say `import` and the library name. For Pandas, is it common to name the library `pd` so that when we reference a function from the Pandas library, we only write `pd` -- not `pandas`.

In [1]:
import pandas as pd

## Reading in Data

Pandas dramatically simplifies the process of reading in data. When we say "reading in data," we mean loading a file into our machine's memory.

When you have a CSV, for example, and then you double-click to open it in Microsoft Excel, the open file is "read into memory." You can now manipulate the CSV.

When we read data into memory in Python, we are creating an object. We will soon explore this object. _(And, as an aside, when we have a file that is greater than the size of our computer's memory, this is approaching "Big Data.")_

Because we are working with a CSV, we will use the [read CSV](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) method. (Note: reading other formats are supported)

In [2]:
# pd refers to "Pandas" (specific encoding for this file -- non-essential for others)
liq = pd.read_csv("../data/iowa_liquor_may_17_18.csv", encoding='cp1252')

*Documentation Pause*

How did we know how to use `pd.read_csv`? Let's take a look at the [documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html). Note the first argument required (`filepath`).
> Take a moment to dissect other arguments and options when reading in data.

We have just created a data structure called a `DataFrame`. See?

In [3]:
type(liq)

pandas.core.frame.DataFrame

## Inspecting our DataFrame: The basics

We'll now perform basic operations on the DataFrame, denoted with comments.

In [4]:
# print the first and last 30 rows
liq

Date  Store Number                         Store Name  \
0        5/2/17          5286                              Sauce   
1        5/1/17          4169            Super Quick 2 / Hubbell   
2        5/1/17          2641  Hy-Vee Drugstore / Council Bluffs   
3        5/1/17          2641  Hy-Vee Drugstore / Council Bluffs   
4        5/3/17          2565        Hy-Vee Food Store / Spencer   
5        5/3/17          5105             Three Brothers Liquors   
6        5/1/17          4301                   Sahota Food Mart   
7        5/1/17          4301                   Sahota Food Mart   
8        5/1/17          4301                   Sahota Food Mart   
9        5/1/17          4301                   Sahota Food Mart   
10       5/1/17          4301                   Sahota Food Mart   
11       5/1/17          4301                   Sahota Food Mart   
12       5/1/17          4301                   Sahota Food Mart   
13       5/1/17          4301                   Sahota Food Mart   
14       5/1/17          4301                   Sahota Food Mart   
15       5/1/17          4301                   Sahota Food Mart   
16       5/1/17          4301                   Sahota Food Mart   
17       5/1/17          4301                   Sahota Food Mart   
18       5/1/17          4301                   Sahota Food Mart   
19       5/1/17          4301                   Sahota Food Mart   
20       5/1/17          4617                     Lickety Liquor   
21       5/1/17          4617                     Lickety Liquor   
22       5/1/17          4617                     Lickety Liquor   
23       5/1/17          4617                     Lickety Liquor   
24       5/1/17          4617                     Lickety Liquor   
25       5/1/17          4617                     Lickety Liquor   
26       5/1/17          4617                     Lickety Liquor   
27       5/1/17          4617                     Lickety Liquor   
28       5/1/17          4617                     Lickety Liquor   
29       5/1/17          4617                     Lickety Liquor   
...         ...           ...                                ...   
427893  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427894  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427895  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427896  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427897  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427898  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427899  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427900  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427901  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427902  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427903  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427904  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427905  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427906  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427907  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427908  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427909  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427910  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427911  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427912  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427913  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427914  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427915  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427916  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427917  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427918  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   
427919  5/31/18          2595  Hy-Vee Wine and Spirits / Denison   

In [5]:
# print the first five rows
liq.head()

Date  Store Number                         Store Name            City  \
0  5/2/17          5286                              Sauce       Iowa City   
1  5/1/17          4169            Super Quick 2 / Hubbell      Des Moines   
2  5/1/17          2641  Hy-Vee Drugstore / Council Bluffs  Council Bluffs   
3  5/1/17          2641  Hy-Vee Drugstore / Council Bluffs  Council Bluffs   
4  5/3/17          2565        Hy-Vee Food Store / Spencer         Spencer   

   Zip Code                                     Store Location      County  \
0   52240.0                    108, College\rIowa City 52240\r     JOHNSON   
1   50317.0               1824 Hubbell Ave\rDes Moines 50317\r        POLK   
2   51501.0  757 W Broadway\rCouncil Bluffs 51501\r(41.2616...  POTTAWATTA   
3   51501.0  757 W Broadway\rCouncil Bluffs 51501\r(41.2616...  POTTAWATTA   
4   51301.0  819 N Grand Ave\rSpencer 51301\r(43.145897, -9...        CLAY   

                 Category Name               Vendor Name  Item Number  \
0             Blended Whiskies           Laird & Company        23827   
1            Canadian Whiskies  CONSTELLATION BRANDS INC        11773   
2  American Cordials & Liqueur     SAZERAC NORTH AMERICA        84207   
3  American Cordials & Liqueur     SAZERAC NORTH AMERICA        84197   
4                Mixto Tequila                 LUXCO INC        89448   

      ...      Pack  Bottle Volume (ml) State Bottle Cost State Bottle Retail  \
0     ...        12               1,000            $4.40               $6.60    
1     ...        48                 200            $1.56               $2.34    
2     ...        10                 600            $6.00               $9.00    
3     ...        10                 600            $6.00               $9.00    
4     ...         6               1,750           $12.00              $18.00    

  Bottles Sold  Sale (Dollars) Volume Sold (Liters)  Volume Sold (Gallons)  \
0            1         $79.20                  1.00                   0.26   
1            1        $112.32                  0.20                   0.05   
2            1          $9.00                  0.60                   0.15   
3            1          $9.00                  0.60                   0.15   
4            3         $18.00                  5.25                   1.38   

   is_may_2017  is_may_2018  
0            1            0  
1            1            0  
2            1            0  
3            1            0  
4            1            0  

[5 rows x 21 columns]

Notice that `.head()` is a method (denoted by parantheses), so it takes arguments.

**Class Question:** What do you think changes if we pass a different number `head()` argument?

In [6]:
# try 10 as an argument
liq.head(10)

Date  Store Number                         Store Name            City  \
0  5/2/17          5286                              Sauce       Iowa City   
1  5/1/17          4169            Super Quick 2 / Hubbell      Des Moines   
2  5/1/17          2641  Hy-Vee Drugstore / Council Bluffs  Council Bluffs   
3  5/1/17          2641  Hy-Vee Drugstore / Council Bluffs  Council Bluffs   
4  5/3/17          2565        Hy-Vee Food Store / Spencer         Spencer   
5  5/3/17          5105             Three Brothers Liquors   North Liberty   
6  5/1/17          4301                   Sahota Food Mart      Des Moines   
7  5/1/17          4301                   Sahota Food Mart      Des Moines   
8  5/1/17          4301                   Sahota Food Mart      Des Moines   
9  5/1/17          4301                   Sahota Food Mart      Des Moines   

   Zip Code                                     Store Location      County  \
0   52240.0                    108, College\rIowa City 52240\r     JOHNSON   
1   50317.0               1824 Hubbell Ave\rDes Moines 50317\r        POLK   
2   51501.0  757 W Broadway\rCouncil Bluffs 51501\r(41.2616...  POTTAWATTA   
3   51501.0  757 W Broadway\rCouncil Bluffs 51501\r(41.2616...  POTTAWATTA   
4   51301.0  819 N Grand Ave\rSpencer 51301\r(43.145897, -9...        CLAY   
5   52317.0  585 HIGHWAY 965\rNorth Liberty 52317\r(41.7381...     JOHNSON   
6   50320.0  1805 SE 14th St\rDes Moines 50320\r(41.57222, ...        POLK   
7   50320.0  1805 SE 14th St\rDes Moines 50320\r(41.57222, ...        POLK   
8   50320.0  1805 SE 14th St\rDes Moines 50320\r(41.57222, ...        POLK   
9   50320.0  1805 SE 14th St\rDes Moines 50320\r(41.57222, ...        POLK   

                 Category Name               Vendor Name  Item Number  \
0             Blended Whiskies           Laird & Company        23827   
1            Canadian Whiskies  CONSTELLATION BRANDS INC        11773   
2  American Cordials & Liqueur     SAZERAC NORTH AMERICA        84207   
3  American Cordials & Liqueur     SAZERAC NORTH AMERICA        84197   
4                Mixto Tequila                 LUXCO INC        89448   
5               American Vodka            A V BRANDS INC       937040   
6            American Brandies  CONSTELLATION BRANDS INC        53214   
7              American Vodkas           Laird & Company        35914   
8              Imported Vodkas           BACARDI USA INC        34359   
9              Imported Vodkas           BACARDI USA INC        34423   

      ...      Pack  Bottle Volume (ml) State Bottle Cost State Bottle Retail  \
0     ...        12               1,000            $4.40               $6.60    
1     ...        48                 200            $1.56               $2.34    
2     ...        10                 600            $6.00               $9.00    
3     ...        10                 600            $6.00               $9.00    
4     ...         6               1,750           $12.00              $18.00    
5     ...         6                 750           $21.99              $32.99    
6     ...        24                 375            $3.22               $4.83    
7     ...        24                 375            $1.93               $2.90    
8     ...        12                 200            $5.00               $7.50    
9     ...        12                 375            $9.00              $13.50    

  Bottles Sold  Sale (Dollars) Volume Sold (Liters)  Volume Sold (Gallons)  \
0            1         $79.20                  1.00                   0.26   
1            1        $112.32                  0.20                   0.05   
2            1          $9.00                  0.60                   0.15   
3            1          $9.00                  0.60                   0.15   
4            3         $18.00                  5.25                   1.38   
5            2        $135.00                  1.50                   0.39   
6            1        $115.92                

In [7]:
# print the last five rows
liq.tail()

Date  Store Number                         Store Name     City  \
427918  5/31/18          2595  Hy-Vee Wine and Spirits / Denison  Denison   
427919  5/31/18          2595  Hy-Vee Wine and Spirits / Denison  Denison   
427920  5/31/18          2595  Hy-Vee Wine and Spirits / Denison  Denison   
427921  5/31/18          2595  Hy-Vee Wine and Spirits / Denison  Denison   
427922  5/31/18          2595  Hy-Vee Wine and Spirits / Denison  Denison   

        Zip Code                                     Store Location    County  \
427918   51442.0  1620 4th Ave, South\rDenison 51442\r(42.012395...  CRAWFORD   
427919   51442.0  1620 4th Ave, South\rDenison 51442\r(42.012395...  CRAWFORD   
427920   51442.0  1620 4th Ave, South\rDenison 51442\r(42.012395...  CRAWFORD   
427921   51442.0  1620 4th Ave, South\rDenison 51442\r(42.012395...  CRAWFORD   
427922   51442.0  1620 4th Ave, South\rDenison 51442\r(42.012395...  CRAWFORD   

            Category Name               Vendor Name  Item Number     ...       \
427918    American Vodkas  McCormick Distilling Co.        36908     ...        
427919    American Vodkas      SAZERAC COMPANY  INC        36978     ...        
427920  Iowa Distilleries          Infinium Spirits        27125     ...        
427921   Blended Whiskies           DIAGEO AMERICAS        25607     ...        
427922   Blended Whiskies           DIAGEO AMERICAS        25608     ...        

       Pack  Bottle Volume (ml) State Bottle Cost State Bottle Retail  \
427918    6               1,750            $7.47              $11.21    
427919    6               1,750            $6.92              $10.38    
427920    6                 750           $22.75              $34.13    
427921   12               1,000            $8.00              $12.00    
427922    6               1,750           $11.96              $17.94    

       Bottles Sold  Sale (Dollars) Volume Sold (Liters)  \
427918            1         $67.26                  1.75   
427919            1         $62.28                  1.75   
427920            1        $204.78                  0.75   
427921            1        $144.00                  1.00   
427922            1        $107.64                  1.75   

        Volume Sold (Gallons)  is_may_2017  is_may_2018  
427918                   0.46            0            1  
427919                   0.46            0            1  
427920                   0.19            0            1  
427921                   0.26            0            1  
427922                   0.46            0            1  

[5 rows x 21 columns]

In [8]:
# identify the shape (rows by columns)
liq.shape

(427923, 21)

Wow! hundreds of thousands of rows rows and tens of columns!

In [9]:
# display the index
liq.index

RangeIndex(start=0, stop=427923, step=1)

In [10]:
# print the columns
liq.columns

Index(['Date', 'Store Number', 'Store Name', 'City', 'Zip Code',
       'Store Location', 'County', 'Category Name', 'Vendor Name',
       'Item Number', 'Item Description', 'Pack', 'Bottle Volume (ml)',
       'State Bottle Cost', 'State Bottle Retail', 'Bottles Sold',
       'Sale (Dollars)', 'Volume Sold (Liters)', 'Volume Sold (Gallons)',
       'is_may_2017', 'is_may_2018'],
      dtype='object')

In [11]:
# examine the datatypes of the columns
liq.dtypes

Date                      object
Store Number               int64
Store Name                object
City                      object
Zip Code                 float64
Store Location            object
County                    object
Category Name             object
Vendor Name               object
Item Number                int64
Item Description          object
Pack                       int64
Bottle Volume (ml)        object
State Bottle Cost         object
State Bottle Retail       object
Bottles Sold               int64
Sale (Dollars)            object
Volume Sold (Liters)     float64
Volume Sold (Gallons)    float64
is_may_2017                int64
is_may_2018                int64
dtype: object

**Class Question:** Why do datatypes matter? What operations could we perform on some datatypes that we could not on others? Note the importance of this in checking dataset integrity.

## Selecting a Columns

We can select columns in two ways. Either we treat the column as an attribute of the DataFrame or we index the DataFrame for a specific element (in this case, the element is a column name).

In [12]:
# select the county column
liq['County']

0            JOHNSON
1               POLK
2         POTTAWATTA
3         POTTAWATTA
4               CLAY
5            JOHNSON
6               POLK
7               POLK
8               POLK
9               POLK
10              POLK
11              POLK
12              POLK
13              POLK
14              POLK
15              POLK
16              POLK
17              POLK
18              POLK
19              POLK
20              POLK
21              POLK
22              POLK
23              POLK
24              POLK
25              POLK
26              POLK
27              POLK
28              POLK
29              POLK
             ...    
427893      CRAWFORD
427894      CRAWFORD
427895      CRAWFORD
427896      CRAWFORD
427897      CRAWFORD
427898      CRAWFORD
427899      CRAWFORD
427900      CRAWFORD
427901      CRAWFORD
427902      CRAWFORD
427903      CRAWFORD
427904      CRAWFORD
427905      CRAWFORD
427906      CRAWFORD
427907      CRAWFORD
427908      CRAWFORD
427909      C

In [13]:
# this is equivalent
liq.County

0            JOHNSON
1               POLK
2         POTTAWATTA
3         POTTAWATTA
4               CLAY
5            JOHNSON
6               POLK
7               POLK
8               POLK
9               POLK
10              POLK
11              POLK
12              POLK
13              POLK
14              POLK
15              POLK
16              POLK
17              POLK
18              POLK
19              POLK
20              POLK
21              POLK
22              POLK
23              POLK
24              POLK
25              POLK
26              POLK
27              POLK
28              POLK
29              POLK
             ...    
427893      CRAWFORD
427894      CRAWFORD
427895      CRAWFORD
427896      CRAWFORD
427897      CRAWFORD
427898      CRAWFORD
427899      CRAWFORD
427900      CRAWFORD
427901      CRAWFORD
427902      CRAWFORD
427903      CRAWFORD
427904      CRAWFORD
427905      CRAWFORD
427906      CRAWFORD
427907      CRAWFORD
427908      CRAWFORD
427909      C

**Class Question:** What if we wanted to select a column that has a space in it? Which method from the above two would we use? Why?

## Renaming Columns

Perhaps we want to rename our columns, like replacing spaces with underscores. There's a few options for doing this.

In [14]:
# rename one or more columns with a dictionary. Note: inplace = True
liq.rename(columns={'Store Number': 'Store_Number', 'Store Name':'Store_Name'}, inplace=True)

In [15]:
# check the result
liq.head()

Date  Store_Number                         Store_Name            City  \
0  5/2/17          5286                              Sauce       Iowa City   
1  5/1/17          4169            Super Quick 2 / Hubbell      Des Moines   
2  5/1/17          2641  Hy-Vee Drugstore / Council Bluffs  Council Bluffs   
3  5/1/17          2641  Hy-Vee Drugstore / Council Bluffs  Council Bluffs   
4  5/3/17          2565        Hy-Vee Food Store / Spencer         Spencer   

   Zip Code                                     Store Location      County  \
0   52240.0                    108, College\rIowa City 52240\r     JOHNSON   
1   50317.0               1824 Hubbell Ave\rDes Moines 50317\r        POLK   
2   51501.0  757 W Broadway\rCouncil Bluffs 51501\r(41.2616...  POTTAWATTA   
3   51501.0  757 W Broadway\rCouncil Bluffs 51501\r(41.2616...  POTTAWATTA   
4   51301.0  819 N Grand Ave\rSpencer 51301\r(43.145897, -9...        CLAY   

                 Category Name               Vendor Name  Item Number  \
0             Blended Whiskies           Laird & Company        23827   
1            Canadian Whiskies  CONSTELLATION BRANDS INC        11773   
2  American Cordials & Liqueur     SAZERAC NORTH AMERICA        84207   
3  American Cordials & Liqueur     SAZERAC NORTH AMERICA        84197   
4                Mixto Tequila                 LUXCO INC        89448   

      ...      Pack  Bottle Volume (ml) State Bottle Cost State Bottle Retail  \
0     ...        12               1,000            $4.40               $6.60    
1     ...        48                 200            $1.56               $2.34    
2     ...        10                 600            $6.00               $9.00    
3     ...        10                 600            $6.00               $9.00    
4     ...         6               1,750           $12.00              $18.00    

  Bottles Sold  Sale (Dollars) Volume Sold (Liters)  Volume Sold (Gallons)  \
0            1         $79.20                  1.00                   0.26   
1            1        $112.32                  0.20                   0.05   
2            1          $9.00                  0.60                   0.15   
3            1          $9.00                  0.60                   0.15   
4            3         $18.00                  5.25                   1.38   

   is_may_2017  is_may_2018  
0            1            0  
1            1            0  
2            1            0  
3            1            0  
4            1            0  

[5 rows x 21 columns]

Bulk rename columns with a list of new column names.

In [16]:
liq.columns

Index(['Date', 'Store_Number', 'Store_Name', 'City', 'Zip Code',
       'Store Location', 'County', 'Category Name', 'Vendor Name',
       'Item Number', 'Item Description', 'Pack', 'Bottle Volume (ml)',
       'State Bottle Cost', 'State Bottle Retail', 'Bottles Sold',
       'Sale (Dollars)', 'Volume Sold (Liters)', 'Volume Sold (Gallons)',
       'is_may_2017', 'is_may_2018'],
      dtype='object')

In [17]:
# declare a list of strings - these strings will become the new column names
cols = ['date', 'store_number', 'store_name', 'city', 
        'zip_code', 'location', 'county', 'category_name',
        'vendor_name', 'item_number', 'item_description', 'pack', 
       'bottle_vol_ml', 'state_bottle_cost', 'state_bottle_retail', 'bottles_sold',
       'sale', 'volumne_sold_l', 'volume_sold_gal', 'is_may_2017', 'is_may_2018']

In [18]:
# use that list to rename our columns - inplace by default
liq.columns = cols

In [19]:
# check out the result!
liq.head()

date  store_number                         store_name            city  \
0  5/2/17          5286                              Sauce       Iowa City   
1  5/1/17          4169            Super Quick 2 / Hubbell      Des Moines   
2  5/1/17          2641  Hy-Vee Drugstore / Council Bluffs  Council Bluffs   
3  5/1/17          2641  Hy-Vee Drugstore / Council Bluffs  Council Bluffs   
4  5/3/17          2565        Hy-Vee Food Store / Spencer         Spencer   

   zip_code                                           location      county  \
0   52240.0                    108, College\rIowa City 52240\r     JOHNSON   
1   50317.0               1824 Hubbell Ave\rDes Moines 50317\r        POLK   
2   51501.0  757 W Broadway\rCouncil Bluffs 51501\r(41.2616...  POTTAWATTA   
3   51501.0  757 W Broadway\rCouncil Bluffs 51501\r(41.2616...  POTTAWATTA   
4   51301.0  819 N Grand Ave\rSpencer 51301\r(43.145897, -9...        CLAY   

                 category_name               vendor_name  item_number  \
0             Blended Whiskies           Laird & Company        23827   
1            Canadian Whiskies  CONSTELLATION BRANDS INC        11773   
2  American Cordials & Liqueur     SAZERAC NORTH AMERICA        84207   
3  American Cordials & Liqueur     SAZERAC NORTH AMERICA        84197   
4                Mixto Tequila                 LUXCO INC        89448   

      ...      pack  bottle_vol_ml state_bottle_cost state_bottle_retail  \
0     ...        12          1,000            $4.40               $6.60    
1     ...        48            200            $1.56               $2.34    
2     ...        10            600            $6.00               $9.00    
3     ...        10            600            $6.00               $9.00    
4     ...         6          1,750           $12.00              $18.00    

  bottles_sold      sale volumne_sold_l  volume_sold_gal  is_may_2017  \
0            1   $79.20            1.00             0.26            1   
1            1  $112.32            0.20             0.05            1   
2            1    $9.00            0.60             0.15            1   
3            1    $9.00            0.60             0.15            1   
4            3   $18.00            5.25             1.38            1   

   is_may_2018  
0            0  
1            0  
2            0  
3            0  
4            0  

[5 rows x 21 columns]

## Notable Column Operations

While this is non-comprehensive, these are a few key column-specific data checks.


**Five-number summary:**  the minimum, first quartile, median, third quartile, and maximum.

(And more! The mean too.)

Five Number Summary (all assumes numeric data):
- **Min:** The smallest value in the column
- **Max:** The largest value in the column
- **Quartile:** A quartile is one fourth of our data
    - **First quartile:** This is the bottom most 25 percent
    - **Median:** The middle value. (Line all values biggest to smallest - median is the middle!) Also the 50th percentile
    - **Third quartile:** This the the top 75 percentile of our data


![](https://www.mathsisfun.com/data/images/quartiles-a.svg)

In [20]:
# note - describe *default* only checks numeric datatypes
liq.describe()

store_number       zip_code    item_number           pack  \
count  427923.000000  427909.000000  427923.000000  427923.000000   
mean     3789.924475   51275.466083   46367.248790      12.423403   
std      1090.596072     986.858053   52774.045141       7.849579   
min      2106.000000   50002.000000     139.000000       1.000000   
25%      2616.000000   50317.000000   27125.000000       6.000000   
50%      3849.000000   51104.000000   38177.000000      12.000000   
75%      4802.000000   52314.000000   64762.000000      12.000000   
max      9937.000000   56201.000000  998546.000000      48.000000   

        bottles_sold  volumne_sold_l  volume_sold_gal    is_may_2017  \
count  427923.000000   427923.000000    427923.000000  427923.000000   
mean        2.251461        1.984745         0.518811       0.489308   
std         3.976579        6.175569         1.631609       0.499886   
min         0.000000        0.010000         0.000000       0.000000   
25%         1.000000        0.750000         0.190000       0.000000   
50%         1.000000        1.500000         0.390000       0.000000   
75%         3.000000        2.000000         0.520000       1.000000   
max       315.000000      551.250000       145.620000       1.000000   

         is_may_2018  
count  427923.000000  
mean        0.510692  
std         0.499886  
min         0.000000  
25%         0.000000  
50%         1.000000  
75%         1.000000  
max         1.000000

> To yourself: consider the **skew** of your series. When does the mean differ significantly from the median? Why does this matter?

In [21]:
# enrichment
# liq.describe(include='all')

**Value Counts:** count the occurrence of each value within our series.

In [22]:
liq.columns

Index(['date', 'store_number', 'store_name', 'city', 'zip_code', 'location',
       'county', 'category_name', 'vendor_name', 'item_number',
       'item_description', 'pack', 'bottle_vol_ml', 'state_bottle_cost',
       'state_bottle_retail', 'bottles_sold', 'sale', 'volumne_sold_l',
       'volume_sold_gal', 'is_may_2017', 'is_may_2018'],
      dtype='object')

In [23]:
# show the most frequent store purchasers
liq.store_name.value_counts()

Hy-Vee #3 / BDI / Des Moines                 3946
Central City 2                               3244
Hy-Vee Food Store / Cedar Falls              3165
Hy-Vee Wine and Spirits / Bettendorf         2646
Hy-Vee Food Store / Coralville               2418
Hy-Vee Food Store / Spencer                  2339
Hy-Vee Food Store #1 / Ames                  2338
Hy-Vee Food Store / Muscatine                2284
Hy-Vee Wine and Spirits / Iowa City          2169
Hy-Vee Food Store #5 / Cedar Rapids          2152
Cyclone Liquors                              2108
Central City Liquor, Inc.                    2056
Hy-Vee Food Store #1 / Mason City            2051
Hy-Vee Food Store #3 / Sioux City            2007
Hy-Vee #7 / Cedar Rapids                     1987
Hy-Vee Wine and Spirits / WDM                1984
Iowa Street Market, Inc.                     1972
Hy-Vee / Waukee                              1942
Hy-Vee Food Store #3 / Cedar Rapids          1932
Hy-Vee Food and Drug / Clinton               1918


In [24]:
# same thing - but only top 10 using splicing we learned in days 1-3
liq.store_name.value_counts()[:10]

Hy-Vee #3 / BDI / Des Moines            3946
Central City 2                          3244
Hy-Vee Food Store / Cedar Falls         3165
Hy-Vee Wine and Spirits / Bettendorf    2646
Hy-Vee Food Store / Coralville          2418
Hy-Vee Food Store / Spencer             2339
Hy-Vee Food Store #1 / Ames             2338
Hy-Vee Food Store / Muscatine           2284
Hy-Vee Wine and Spirits / Iowa City     2169
Hy-Vee Food Store #5 / Cedar Rapids     2152
Name: store_name, dtype: int64

**Unique values:** Determine the number of distinct values within a given series.

In [25]:
# how many distinct stores are there in Iowa?
liq.store_name.nunique()

1583

In [26]:
# what are those stores called? (notice: too many to print them all!)
liq.store_name.unique()

array(['Sauce', 'Super Quick 2 / Hubbell',
       'Hy-Vee Drugstore / Council Bluffs', ...,
       'Liberty View Wine and Spirits',
       "Casey's General Store #1901 / Des Moines",
       'CVS Pharmacy #10162 / Des Moines'], dtype=object)

## Filtering

Filtering and sorting are key processes that allow us to drill into the 'nitty gritty' and cross sections of our dataset.

To filter, we use a process called **Boolean Filtering**, wherein we define a Boolean condition, and use that Boolean condition to filer on our DataFrame.

Recall: our given dataset has a column `is_may_2017` when a given sale occurred in the month of May 2017. The column is equal to 1 when the sale did occur in May 2017, and zero otherwise (which means it occurred in May 2018 as those are the only two months in our dataset).

Let's calculate the **average pack size** for sales in May 2017.

> Think: What are the component parts of this problem?

In [27]:
# First, create a Boolean filter for is_may_2017
liq.is_may_2017 == 1

0          True
1          True
2          True
3          True
4          True
5          True
6          True
7          True
8          True
9          True
10         True
11         True
12         True
13         True
14         True
15         True
16         True
17         True
18         True
19         True
20         True
21         True
22         True
23         True
24         True
25         True
26         True
27         True
28         True
29         True
          ...  
427893    False
427894    False
427895    False
427896    False
427897    False
427898    False
427899    False
427900    False
427901    False
427902    False
427903    False
427904    False
427905    False
427906    False
427907    False
427908    False
427909    False
427910    False
427911    False
427912    False
427913    False
427914    False
427915    False
427916    False
427917    False
427918    False
427919    False
427920    False
427921    False
427922    False
Name: is_may_2017, Lengt

In [28]:
# use that Boolean filter to index the DataFrame, only printing rows where True
liq[liq.is_may_2017 == 1]

date  store_number                         store_name  \
0        5/2/17          5286                              Sauce   
1        5/1/17          4169            Super Quick 2 / Hubbell   
2        5/1/17          2641  Hy-Vee Drugstore / Council Bluffs   
3        5/1/17          2641  Hy-Vee Drugstore / Council Bluffs   
4        5/3/17          2565        Hy-Vee Food Store / Spencer   
5        5/3/17          5105             Three Brothers Liquors   
6        5/1/17          4301                   Sahota Food Mart   
7        5/1/17          4301                   Sahota Food Mart   
8        5/1/17          4301                   Sahota Food Mart   
9        5/1/17          4301                   Sahota Food Mart   
10       5/1/17          4301                   Sahota Food Mart   
11       5/1/17          4301                   Sahota Food Mart   
12       5/1/17          4301                   Sahota Food Mart   
13       5/1/17          4301                   Sahota Food Mart   
14       5/1/17          4301                   Sahota Food Mart   
15       5/1/17          4301                   Sahota Food Mart   
16       5/1/17          4301                   Sahota Food Mart   
17       5/1/17          4301                   Sahota Food Mart   
18       5/1/17          4301                   Sahota Food Mart   
19       5/1/17          4301                   Sahota Food Mart   
20       5/1/17          4617                     Lickety Liquor   
21       5/1/17          4617                     Lickety Liquor   
22       5/1/17          4617                     Lickety Liquor   
23       5/1/17          4617                     Lickety Liquor   
24       5/1/17          4617                     Lickety Liquor   
25       5/1/17          4617                     Lickety Liquor   
26       5/1/17          4617                     Lickety Liquor   
27       5/1/17          4617                     Lickety Liquor   
28       5/1/17          4617                     Lickety Liquor   
29       5/1/17          4617                     Lickety Liquor   
...         ...           ...                                ...   
209356  5/31/17          2670     Hy-Vee Food Store / Coralville   
209357  5/31/17          2670     Hy-Vee Food Store / Coralville   
209358  5/31/17          2670     Hy-Vee Food Store / Coralville   
209359  5/31/17          2670     Hy-Vee Food Store / Coralville   
209360  5/31/17          2670     Hy-Vee Food Store / Coralville   
209361  5/31/17          2670     Hy-Vee Food Store / Coralville   
209362  5/31/17          2670     Hy-Vee Food Store / Coralville   
209363  5/31/17          2670     Hy-Vee Food Store / Coralville   
209364  5/31/17          2670     Hy-Vee Food Store / Coralville   
209365  5/31/17          2670     Hy-Vee Food Store / Coralville   
209366  5/31/17          2670     Hy-Vee Food Store / Coralville   
209367  5/31/17          2670     Hy-Vee Food Store / Coralville   
209368  5/31/17          2670     Hy-Vee Food Store / Coralville   
209369  5/31/17          2670     Hy-Vee Food Store / Coralville   
209370  5/31/17          2670     Hy-Vee Food Store / Coralville   
209371  5/31/17          2670     Hy-Vee Food Store / Coralville   
209372  5/31/17          2670     Hy-Vee Food Store / Coralville   
209373  5/31/17          2670     Hy-Vee Food Store / Coralville   
209374  5/31/17          2670     Hy-Vee Food Store / Coralville   
209375  5/31/17          2670     Hy-Vee Food Store / Coralville   
209376  5/31/17          2670     Hy-Vee Food Store / Coralville   
209377  5/31/17          2670     Hy-Vee Food Store / Coralville   
209378  5/31/17          2670     Hy-Vee Food Store / Coralville   
209379  5/31/17          2670     Hy-Vee Food Store / Coralville   
209380  5/31/17          2670     Hy-Vee Food Store / Coralville   
209381  5/31/17          2670     Hy-Vee Food Store / Coralville   
209382  5/31/17          2670     Hy-Vee Food Store / Coralville   

Great! This DataFrame is only sales in May 2017!

Now, we need to calculate the average pack size of this DataFrame.

In [29]:
# get the pack column data...
liq[liq.is_may_2017 == 1].pack

0         12
1         48
2         10
3         10
4          6
5          6
6         24
7         24
8         12
9         12
10        24
11        12
12         6
13        24
14        12
15        12
16         6
17        24
18        12
19         6
20        12
21        12
22        12
23        12
24        12
25        12
26        12
27        12
28        48
29        24
          ..
209356     6
209357     6
209358    12
209359    12
209360    12
209361    12
209362     6
209363    12
209364    12
209365    12
209366    12
209367    12
209368    12
209369     6
209370     6
209371    12
209372    12
209373     6
209374    12
209375    12
209376    12
209377    12
209378    12
209379    12
209380    12
209381    12
209382    12
209383    12
209384    12
209385    12
Name: pack, Length: 209386, dtype: int64

In [30]:
# now calculate the average!
liq[liq.is_may_2017 == 1].pack.describe()

count    209386.000000
mean         12.414278
std           7.754952
min           1.000000
25%           6.000000
50%          12.000000
75%          12.000000
max          48.000000
Name: pack, dtype: float64

**The average pack size for sales in May 2017 is 12.41!**

## Sorting

We can sort one column of our DataFrame as well.

In [31]:
# let's sort by largest bottles in a single day
liq.sort_values(by='bottles_sold', ascending = False)

date  store_number                            store_name  \
285425  5/10/18          3814           Costco Wholesale #788 / WDM   
256127   5/8/18          4677   Costco Wholesale #1111 / Coralville   
256128   5/8/18          4677   Costco Wholesale #1111 / Coralville   
136200  5/22/17          3814                 Costco Wholesale #788   
285424  5/10/18          3814           Costco Wholesale #788 / WDM   
136197  5/22/17          3814                 Costco Wholesale #788   
100130  5/16/17          4677   Costco Wholesale #1111 / Coralville   
38373    5/4/17          2593           Hy-Vee Food Store / Carroll   
223994   5/2/18          4319         Fareway Stores #703 / Humbolt   
76245   5/11/17          3420     Sam's Club 6344 / Windsor Heights   
176134  5/25/17          2593           Hy-Vee Food Store / Carroll   
153021  5/23/17          5431    Brew Coffee Wine Spirit and Cigars   
217921   5/1/18          4255     Fareway Stores #058 / Orange City   
217824   5/1/18          5151                            IDA Liquor   
54675    5/8/17          2650      Hy-Vee Wine and Spirits / Harlan   
161308  5/24/17          4254      Fareway Stores #849 / Emmetsburg   
278571   5/9/18          4162    Fareway Stores #963 / Cedar Rapids   
67638   5/10/17          3524                Sam's Club 6568 / Ames   
219766   5/1/18          5431    Brew Coffee Wine Spirit and Cigars   
389772  5/25/18          4051       Fareway Stores #639 / Maquoketa   
293798  5/14/18          4285     Fareway Stores #922 / New Hampton   
34547    5/4/17          2517         Hy-Vee Food Store #1 / Newton   
159504  5/24/17          4311      Fareway Stores #508 / Fort Dodge   
256305   5/8/18          3385        Sam's Club 8162 / Cedar Rapids   
113730  5/17/17          4228          Fareway Stores #462 / Vinton   
176558  5/25/17          4320           Fareway Stores #470 / Perry   
231757   5/3/18          3420     Sam's Club 6344 / Windsor Heights   
270548   5/9/18          4311      Fareway Stores #508 / Fort Dodge   
75047   5/11/17          3814                 Costco Wholesale #788   
18030    5/2/17          2576  Hy-Vee Wine and Spirits / Storm Lake   
...         ...           ...                                   ...   
81705   5/11/17          2633          Hy-Vee #3 / BDI / Des Moines   
170602  5/25/17          2515     Hy-Vee Food Store #1 / Mason City   
189815  5/30/17          5318              First Stop Beverage Shop   
218873   5/1/18          5174          Jim's Food   /  Sullivan Ave   
114905  5/17/17          2536                HY-VEE / PLEASANT HILL   
102013  5/16/17          2590   Hy-Vee Food Store #5 / Cedar Rapids   
288381  5/11/18          2626   Hy-Vee Drugstore / University / DSM   
99997   5/16/17          2190             Central City Liquor, Inc.   
288378  5/11/18          2626   Hy-Vee Drugstore / University / DSM   
368772  5/23/18          2501                     Hy-Vee  #2 / Ames   
103116  5/16/17          2513      Hy-Vee Food Store #2 / Iowa City   
256911   5/8/18          3644               Wal-Mart 2764 / Altoona   
363345  5/23/18          2616        Hy-Vee Food and Drug / Clinton   
281612  5/10/18          4901       Fareway Stores #137 / Polk City   
248139   5/7/18          2560            Hy-Vee Food Store / Marion   
36909    5/4/17          2633          Hy-Vee #3 / BDI / Des Moines   
260568   5/8/18          3632               Wal-Mart 2004 / Dubuque   
33916    5/4/17          3670            Wal-Mart 2827 / Coralville   
218836   5/1/18          3993  New Star Liquor / W 4th S / Waterloo   
75319   5/11/17          3731             Wal-Mart 1241 / Davenport   
217124   5/1/18          5274                  FCA Kingsley C-Store   
108233  5/16/17          2597          Hy-Vee Food Store / Cherokee   
73489   5/10/17          2508   Hy-Vee Food Store #1 / Cedar Rapids   
232884   5/3/18          3354           Sam's Club 8238 / Davenport   
126876  5/18/17          5416      

In [32]:
# let's sort by largest bottles in a single day - but just the top 10
liq.sort_values(by='bottles_sold', ascending = False).head(10)

date  store_number                           store_name  \
285425  5/10/18          3814          Costco Wholesale #788 / WDM   
256127   5/8/18          4677  Costco Wholesale #1111 / Coralville   
256128   5/8/18          4677  Costco Wholesale #1111 / Coralville   
136200  5/22/17          3814                Costco Wholesale #788   
285424  5/10/18          3814          Costco Wholesale #788 / WDM   
136197  5/22/17          3814                Costco Wholesale #788   
100130  5/16/17          4677  Costco Wholesale #1111 / Coralville   
38373    5/4/17          2593          Hy-Vee Food Store / Carroll   
223994   5/2/18          4319        Fareway Stores #703 / Humbolt   
76245   5/11/17          3420    Sam's Club 6344 / Windsor Heights   

                   city  zip_code  \
285425  West Des Moines   50266.0   
256127       Coralville   52241.0   
256128       Coralville   52241.0   
136200  West Des Moines   50266.0   
285424  West Des Moines   50266.0   
136197  West Des Moines   50266.0   
100130       Coralville   52241.0   
38373           Carroll   51401.0   
223994         Humboldt   50548.0   
76245   Windsor Heights   50311.0   

                                                 location    county  \
285425  7205 Mills Civic Pkwy\rWest Des Moines 50266\r...    Dallas   
256127  2900 Heartland Dr\rCoralville 52241\r(41.69802...   JOHNSON   
256128  2900 Heartland Dr\rCoralville 52241\r(41.69802...   JOHNSON   
136200  7205 Mills Civic Pkwy\rWest Des Moines 50266\r...    Dallas   
285424  7205 Mills Civic Pkwy\rWest Des Moines 50266\r...    Dallas   
136197  7205 Mills Civic Pkwy\rWest Des Moines 50266\r...    Dallas   
100130  2900 Heartland Dr\rCoralville 52241\r(41.69802...   JOHNSON   
38373   905 US Highway 30 West\rCarroll 51401\r(42.070...   CARROLL   
223994  1700 10th Avenue N\rHumboldt 50548\r(42.731879...  HUMBOLDT   
76245   1101 73RD STREET\rWindsor Heights 50311\r(41.5...      Polk   

             category_name               vendor_name  item_number  \
285425     American Vodkas       Levecke Corporation       936600   
256127      Cocktails /RTD       Levecke Corporation       962094   
256128     American Vodkas       Levecke Corporation       936600   
136200      Imported Vodka          MISA Imports Inc       987514   
285424      Cocktails /RTD       Levecke Corporation       962094   
136197  100% Agave Tequila          MISA Imports Inc       989289   
100130   Canadian Whiskies  CONSTELLATION BRANDS INC        11788   
38373    Canadian Whiskies  CONSTELLATION BRANDS INC        11788   
223994   Canadian Whiskies  CONSTELLATION BRANDS INC        11788   
76245    Canadian Whiskies  CONSTELLATION BRANDS INC        11788   

           ...      pack  bottle_vol_ml state_bottle_cost state_bottle_retail  \
285425     ...         6          1,750            $8.93              $13.40    
256127     ...         6          1,750            $6.00               $9.00    
256128     ...         6          1,750            $8.93              $13.40    
136200     ...         6          1,750           $14.48              $21.72    
285424     ...         6          1,750            $6.00               $9.00    
136197     ...         6          1,750           $14.90              $22.35    
100130     ...         6          1,750           $10.45              $15.68    
38373      ...         6          1,750           $10.45              $15.68    
223994     ...         6          1,750           $10.45              $15.68    
76245      ...         6          1,750           $10.45              $15.68    

       bottles_sold      sale volumne_sold_l  volume_sold_gal  is_may_2017  \
285425          315   $80.40          551.25           145.62            0   
256127          270   $54.00          472.50           124.82            0   
256128          270   $80.40          472.50           124.82            0   
136200          240  $130.32          420.00           110.95            1   
285424  

## Independent Exercises

Do your best to complete the following prompts. Don't hesitate to look at code we wrote together!

Print the first 15 rows of the whole DataFrame.

In [33]:
# your answer here
liq.head(15)

date  store_number                         store_name            city  \
0   5/2/17          5286                              Sauce       Iowa City   
1   5/1/17          4169            Super Quick 2 / Hubbell      Des Moines   
2   5/1/17          2641  Hy-Vee Drugstore / Council Bluffs  Council Bluffs   
3   5/1/17          2641  Hy-Vee Drugstore / Council Bluffs  Council Bluffs   
4   5/3/17          2565        Hy-Vee Food Store / Spencer         Spencer   
5   5/3/17          5105             Three Brothers Liquors   North Liberty   
6   5/1/17          4301                   Sahota Food Mart      Des Moines   
7   5/1/17          4301                   Sahota Food Mart      Des Moines   
8   5/1/17          4301                   Sahota Food Mart      Des Moines   
9   5/1/17          4301                   Sahota Food Mart      Des Moines   
10  5/1/17          4301                   Sahota Food Mart      Des Moines   
11  5/1/17          4301                   Sahota Food Mart      Des Moines   
12  5/1/17          4301                   Sahota Food Mart      Des Moines   
13  5/1/17          4301                   Sahota Food Mart      Des Moines   
14  5/1/17          4301                   Sahota Food Mart      Des Moines   

    zip_code                                           location      county  \
0    52240.0                    108, College\rIowa City 52240\r     JOHNSON   
1    50317.0               1824 Hubbell Ave\rDes Moines 50317\r        POLK   
2    51501.0  757 W Broadway\rCouncil Bluffs 51501\r(41.2616...  POTTAWATTA   
3    51501.0  757 W Broadway\rCouncil Bluffs 51501\r(41.2616...  POTTAWATTA   
4    51301.0  819 N Grand Ave\rSpencer 51301\r(43.145897, -9...        CLAY   
5    52317.0  585 HIGHWAY 965\rNorth Liberty 52317\r(41.7381...     JOHNSON   
6    50320.0  1805 SE 14th St\rDes Moines 50320\r(41.57222, ...        POLK   
7    50320.0  1805 SE 14th St\rDes Moines 50320\r(41.57222, ...        POLK   
8    50320.0  1805 SE 14th St\rDes Moines 50320\r(41.57222, ...        POLK   
9    50320.0  1805 SE 14th St\rDes Moines 50320\r(41.57222, ...        POLK   
10   50320.0  1805 SE 14th St\rDes Moines 50320\r(41.57222, ...        POLK   
11   50320.0  1805 SE 14th St\rDes Moines 50320\r(41.57222, ...        POLK   
12   50320.0  1805 SE 14th St\rDes Moines 50320\r(41.57222, ...        POLK   
13   50320.0  1805 SE 14th St\rDes Moines 50320\r(41.57222, ...        POLK   
14   50320.0  1805 SE 14th St\rDes Moines 50320\r(41.57222, ...        POLK   

                  category_name               vendor_name  item_number  \
0              Blended Whiskies           Laird & Company        23827   
1             Canadian Whiskies  CONSTELLATION BRANDS INC        11773   
2   American Cordials & Liqueur     SAZERAC NORTH AMERICA        84207   
3   American Cordials & Liqueur     SAZERAC NORTH AMERICA        84197   
4                 Mixto Tequila                 LUXCO INC        89448   
5                American Vodka            A V BRANDS INC       937040   
6             American Brandies  CONSTELLATION BRANDS INC        53214   
7               American Vodkas           Laird & Company        35914   
8               Imported Vodkas           BACARDI USA INC        34359   
9               Imported Vodkas           BACARDI USA INC        34423   
10            Imported Brandies           Jim Beam Brands        47790   
11            Imported Brandies    REMY COINTREAU USA INC        49084   
12              Whiskey Liqueur      SAZERAC COMPANY  INC        64868   
13            Canadian Whiskies           DIAGEO AMERICAS        10805   
14           100% Agave Tequila           Jim Beam Brands        89836   

       ...      pack  bottle_vol_ml state_bottle_cost state_bottle_retail  \
0      ...        12          1,000            $4.40               $6.60    
1      ...        48            200            $1.56               $2.34    
2      ...        10            600            $6.00        

Identify the average number of bottles sold.

In [34]:
# your answer here
liq.bottles_sold.describe()

count    427923.000000
mean          2.251461
std           3.976579
min           0.000000
25%           1.000000
50%           1.000000
75%           3.000000
max         315.000000
Name: bottles_sold, dtype: float64

In [35]:
# option 2
liq.bottles_sold.mean()

2.251461127352351

Identify unique number of pack sizes.

In [36]:
# your answer here
liq.pack.nunique()

15

Determine the average volume sold (in gallons) in May 2018.

In [37]:
# your answer here
liq[liq.is_may_2018 == 1].volume_sold_gal.describe()

count    218537.000000
mean          0.521378
std           1.612147
min           0.000000
25%           0.190000
50%           0.390000
75%           0.520000
max         145.620000
Name: volume_sold_gal, dtype: float64

**Challenge:** Identify the top store (as measured by `bottles_sold`) in Johnson County.

In [38]:
# your answer here
liq[liq.county == 'JOHNSON'].sort_values(by='bottles_sold', ascending = False).head(1)

date  store_number                           store_name        city  \
256128  5/8/18          4677  Costco Wholesale #1111 / Coralville  Coralville   

        zip_code                                           location   county  \
256128   52241.0  2900 Heartland Dr\rCoralville 52241\r(41.69802...  JOHNSON   

          category_name          vendor_name  item_number     ...      pack  \
256128  American Vodkas  Levecke Corporation       936600     ...         6   

        bottle_vol_ml state_bottle_cost state_bottle_retail bottles_sold  \
256128          1,750            $8.93              $13.40           270   

           sale volumne_sold_l  volume_sold_gal  is_may_2017  is_may_2018  
256128  $80.40           472.5           124.82            0            1  

[1 rows x 21 columns]

## Recap

We covered a lot of ground! It's ok if this takes a while to gel.

```python

# basic DataFrame operations
df.head()
df.tail()
df.shape
df.columns
df.Index

# selecting columns
df.column_name
df['column_name']

# renaming columns
df.rename({'old_name':'new_name'}, inplace=True)
df.columns = ['new_column_a', 'new_column_b']

# notable columns operations
df.describe() # five number summary
df.column_name.nunique() # number of unique values
df.column_name.value_counts() # number of occurrences of each value in column

# filtering
df[df.column_name < 50] # filter column to be less than 50

# sorting
df.sort_values(by='column_name', ascending = False) # sort biggest to smallest

```


It's common to refer back to your own code *all the time.* Don't hesistate to reference this guide!

Moreover, there's a file `intro-to-pandas-i-answers.py`, that makes referencing this in the future easy. (Or just export this notebook to a `.py` via Files > Download as > .py ) 🐼


